### ISSUE TITLE
Improve display of icons in configurable profile cards

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#3720]
The current way we are adding icons to condition cards is by putting them in their own field. This looks ok on desktop but pretty bad on mobile:

**Desktop:**
![image](https://user-images.githubusercontent.com/1286381/28646226-8bd8ddde-7215-11e7-9cb3-a825c3dd781f.png)

**Mobile:**
![image](https://user-images.githubusercontent.com/1286381/28646236-92d95514-7215-11e7-83d6-627be5edcd92.png)

Make it possible for the icon to appear vertically centered alongside the label and value of the field on both desktop and mobile, as in this mockup:
![image](https://user-images.githubusercontent.com/1286381/28646270-b9a02b32-7215-11e7-8db1-79f5881b0c10.png)

This could be done by making it possible for the configurer to set an icon for a particular field. Regardless, the icon should be conditional, so we could only show it for a high-risk pregnancy and not for a normal pregnancy, for example.

**Comment by garethbowen** [comment]
It should be fairly straightforward to add an `icon` property to the `field` emitted from the configuration. This `icon` can reference one of the configured icons in resources.

**Comment by amandacilek** [comment]
This could be a variation of the form field row type (one version with icon, one without). The icon should be contained within a 30 x 30 bounding box, should be centered vertically within the row, and should be "pinned" to the far right of the row, so that as the row resizes, the icon always stays the same distance from the right edge.

**Comment by garethbowen** [comment]
@amandacilek How does the icon look on desktop? I'm worried if it's right aligned in the first column in a two-column layout it'll look more closely associated with the second column than the first. I can't find any specs or screenshots that document desktop view.

**Comment by amandacilek** [comment]
Hi Gareth,

Hm, you're completely right. My recommendation was based on the assumption that the data is displayed in one column (I didn't look at the top mockups carefully!). I'm curious to know more about the rules behind when we switch to two columns. The potential for two column is only relevant for condition cards and the top bio section, is that right? It looks like the list of People/Family Members, Tasks, and History are always one column. And I agree with that - we don't want to display family members or tasks in a two column layout. 

Looking at an immunization example (attached) I can see why two column might be nice as a space saving measure. At the same time, I'm not sure the intention was ever to have condition cards quite this large. I also think it opens up other questions... what do we do if the text for one of these fields is very long, or has a long notes section? What if we eventually want to make the rows clickable so that you can drill down to a page with more details? I guess I'm leaning towards making the desktop layout of cards always one column, which makes for consistent card rules and also matches mobile. 

I'd like to learn a little more about all the rules we currently have in place for styling condition cards so I can provide a recommendation here.[
<img width="819" alt="screen shot 2017-09-26 at 11 42 56 am" src="https://user-images.githubusercontent.com/26235957/30854037-c8e20522-a2b0-11e7-8dd1-58582951c1a2.png">
](url)

**Comment by sglangevin** [comment]
Personally, I think that having all condition cards be one column on desktop is a waste of space. While I agree that cards like this immunization card are larger than what I think we had intended to create with condition cards, I do think it makes sense to allow up to 2 columns on desktop layouts. If you have all your data left-aligned and an icon all the way on the right, that seems like it's as confusing as it is to have the same situation in 2-column view because the icon is still far away from the information it relates to.

As far as how icons should look on desktop, it could get tricky in cases like the MMRV title in your screenshot above because in that case you might want the icon to appear below the title versus some of the shorter titles in your screenshot (like Typhoid) where you could easily place the icon immediately to the right of the title. Perhaps icons should be part of the second line (e.g. 0 of 2) when there are 2 lines and should fill the whole box (both lines) when they are in an otherwise blank box or when there is only 1 line of information.

What do you think @amandacilek?

**Comment by garethbowen** [comment]
@amandacilek Here is [our documentation](https://github.com/medic/medic-docs/blob/master/configuration/contact-summary.md) about contact summary configuration. This line is particularly relevant:

> [field properties:] `width` is how wide out of 12 the field should be shown. Common values are 12 for full width, 6 for half width, or 3 for quarter width.

This means a 4 column layout is possible, or indeed a 12 column layout. This width is overridden on mobile so it always ends up being 1 column.

I agree with Sharon's idea of having the icon beside the value, but we might have to reduce the icon to 15px (from 30px) to fit within the second line.

I also have some concerns about the horizontal label/value layout because as we need all the values to be aligned we have to fix the width of the labels. This means it doesn't work well when some labels are really long (eg: "MMRV (Merasles, Mumps, Rubella, Varicella)" from the above screenshot), or when the values are really long (eg: the "Notes" field). Without reconfiguration these fields would end up ~ 50% wide and have lots of line wrapping on mobile, eg:

![bad-example](https://user-images.githubusercontent.com/2064938/30880983-bbf98b0e-a360-11e7-93ae-e110345a0d69.png)

Unless you have a quick solution I suggest we stick to the vertical label/value configuration on desktop and mobile until we have time to think these issues through. What do you think?

**Comment by amandacilek** [comment]
In regards to the form label and value being on the same line or not, I am fine having NOT having them on the same line (i.e. layout staying how it is currently). This was never a big deal for me and I may have even said that in the original spec at one point.

For the condition cards: Thanks for sharing the additional info, Gareth. I can't think of a situation where we would want to allow 12 columns of data. If it were up to me I would probably set a smaller limit here, something like max 4 or 6 columns. If we want to keep the option of 12 columns available then I would say let's get rid of the idea of adding an icon altogether.

If the choice is between adding the icon to the first or the second line I might actually prefer adding it before the first line of text. Although, I think an icon at 15px would be quite small and difficult to identify, so if the icon must be that small in order to fit next to the text then I'd rather an alternative solution.

**Comment by garethbowen** [comment]
>  I can't think of a situation where we would want to allow 12 columns of data.

We rely on bootstrap for the widths, and I guess they chose 12 because it divides up well, so you can have 1, 2, 3, 4, 6, or 12 column layouts. This format also allows for mixing the widths so you could have a row with 1, 1, 2, 3, 5 if your data (or your configurer) is crazy enough.

I'm guessing that's how icons are added at the moment. From the initial screenshot you have "completed anc visits" which is probably a 6 wide (6/12 or 50%), then the "status" which is 5 wide, and finally the icon which is 1 wide. On desktop this means they all end up on the same line with the icon in the last twelfth of the row, but on mobile all rows become 100% so the icon gets its own row.

I think the options we have are...

1. Left align the icon inside the column. This makes it clear to what it refers but means the label and value won't line up with other fields. Maybe not a big deal...
2. Right align the icon on mobile, and have the icon below (or above) on desktop. This will make the row taller on desktop but that's probably ok.
3. Make the icon ~ 15px high and have it inline either with the label or value either on the left or right. This may make it hard to see but will render the same on mobile and desktop.
4. Move this issue out of 2.14.0 until we've had a chance to think it through some more.

@sglangevin @amandacilek Any preference? Any other solutions?

**Comment by sglangevin** [comment]
Sorry, I missed this comment somehow @garethbowen. 

> 1. Left align the icon inside the column. This makes it clear to what it refers but means the label and value won't line up with other fields. Maybe not a big deal...

This was my initial thought as well, but I'm concerned it will look odd, especially on mobile. What about left-aligning on desktop and right-aligning on mobile?

> 2. Right align the icon on mobile, and have the icon below (or above) on desktop. This will make the row taller on desktop but that's probably ok.

I could see us putting the icon below on desktop and right-aligning it on mobile.

> 3. Make the icon ~ 15px high and have it inline either with the label or value either on the left or right. This may make it hard to see but will render the same on mobile and desktop.

I know I suggested this, but 15px seems too small for the icon so I think we should scrap this idea.

> 4. Move this issue out of 2.14.0 until we've had a chance to think it through some more.

I'd like to get this done so that we can have a nice-looking profile page, so I'd prefer to keep it in 2.14.0.

So, to summarize, I'd suggest one of the following:
1. Left-align on desktop (icon comes before text) and right-align on mobile. This way the icon can stay at 30px and will be centered vertically on both desktop and mobile.
2. Put the icon below the value on desktop and right-align on mobile. Again, icon stays at 30px, which I think is important.

@amandacilek do you have other ideas? Or could we mock up these two options so we can see how it would look in a situation like the immunization card above?

**Comment by amandacilek** [comment]
Alright jumping back in here... 

A suggestion of mine in relation to the row updates was that ALL individual rows are separated from each other by a 1px grey line. If we have that little grey border separating the rows, and if we extend that to include adding a 1px vertical separation line when we have more than one column of rows, that actually helps quite a bit.

See attached mockup: the border lines clearly divide the content rows and make it clearer which row the icon is meant to be associated with.
![condition icon - 1](https://user-images.githubusercontent.com/26235957/31148872-e242d0fe-a842-11e7-976c-fd8b0bdf4fbb.png)

**Comment by amandacilek** [comment]
I would like to add the border lines (if we haven't already) no matter what.

The lines are still a subtle element though and it's admittedly possible that confusion might still arise if we go with the above icon placement.

Of the other two options mentioned (either left-aligning the icon on desktop, or dropping the icon down to it's own line) I prefer the former. The left-aligned icon does push the text out and break the nice vertical alignment of all the text in the card, but at the same time I feel it's a more elegant solution than having the icon below. I could also imagine a future where we might eventually want
![condition icon - 2](https://user-images.githubusercontent.com/26235957/31149208-fbd28b30-a843-11e7-9b9d-84e36dae4ef1.png)
![condition icon - 3](https://user-images.githubusercontent.com/26235957/31149207-fbcb9c30-a843-11e7-9362-cd288445d07a.png)

 to associate an icon to every single row in a card, in which case it will definitely be preferable to have the icon and text side by side, as we do throughout the rest of the app.

Attaching screenshots showing those two options:

**Comment by sglangevin** [comment]
I actually like the first mock-up you did with the icon right-aligned and including the gray lines between fields. I think that makes the association clearer and the immunization card looks nicer that way, in my opinion. If you guys don't like that one, I'm also fine with the left-aligned icon.

**Comment by amandacilek** [comment]
I agree, I do like that one. @garethbowen does the addition of the grey lines help alleviate your original concern about the icon being mistakenly associated with the wrong text?

**Comment by garethbowen** [comment]
Yeah the grey lines are great - I'm just trying to work out how to implement them (it's not as easy as you'd think).

**Comment by garethbowen** [comment]
The grey lines in between aren't possible as shown in the mockups. The main problem is because the width of each item is configurable there's no way for me to tell which item is at the end of each row.

The closest I got is fairly close but breaks the rounded corners at the bottom of the card.

One workaround is to have some padding between the grid and the card so the card still has rounded borders, eg:

![padding](https://user-images.githubusercontent.com/2064938/31156636-25cfd900-a912-11e7-8875-d4bb2462edcd.png)

This conveys the information but isn't exactly the look you were going for.

Given this we're left to pick between these options...

1. Right align with grey borders and some padding on the left, right, and bottom of the grid (as in the above screenshot).
2. Left-align on desktop (icon comes before text) and right-align on mobile. This way the icon can stay at 30px and will be centered vertically on both desktop and mobile.
3. Put the icon below the value on desktop and right-align on mobile. Again, icon stays at 30px, which I think is important.

@amandacilek Let's talk on Hangouts or Slack tomorrow and make a decision. GitHub is useless for conversation...

**Comment by garethbowen** [comment]
Summary of Slack decision...

> Gareth: Lets go with left aligned desktop and right aligned mobile for now to keep it simple until we get feedback from users. What do you think?
> Amanda: Yeah I'm happy with that. And like I said, that layout works especially well if we decide we want lots of things to have icons.

**Comment by garethbowen** [comment]
Done and documentation updated.

**Comment by sglangevin** [comment]
Looks good - can you center the icon vertically? Right now it looks like it's at the top (this is a 30x30px icon, the 60x60px version looked more centered but too big - see 2nd screenshot):
![image](https://user-images.githubusercontent.com/1286381/31205282-4c19010a-a925-11e7-8bdc-6a05decaa292.png)

![image](https://user-images.githubusercontent.com/1286381/31205314-85d8acb0-a925-11e7-9ebe-de9c540d98c2.png)

**Comment by sglangevin** [comment]
Back to you @garethbowen

**Comment by sglangevin** [comment]
Looks good on desktop but now it's also left-aligned on mobile and it should be right-aligned on mobile. Back to you @garethbowen.

**Comment by garethbowen** [comment]
Fixed.

**Comment by sglangevin** [comment]
LGTM, moving to ready.
